src/audio.cpp

```diff
@@ -0,0 +1,29 @@
+#include "audio.hpp"
+#include "config.h"
+
+#ifdef HAVE_PULSE
+#include "pulse.hpp"
+#endif
+
+#ifdef HAVE_PIPEWIRE
+#include "pipewire.hpp"
+#endif
+
+AudioReader *AudioReader::create(AudioReaderParams params)
+{
+#ifdef HAVE_PIPEWIRE
+    AudioReader *pw = new PipeWireReader;
+    pw->params = params;
+    if (pw->init())
+        return pw;
```

Comment:
This seems to silently choose pipewire by default and use pulse as a fallback. What if both are built-in and both available, but the user wants to select pulse?

Reply:
If user is running pipewire I'd say there is no reason to choose the pulse backend.

Reply:
It could be merely for testing, if pipewire backend isn't working or if the user has a preference.

Reply:
maybe it could be an env var instead of an option to make it easy?

Reply:
On my system the pulse backend is completely broken when using the default output selection (just `-a`), it hangs wf-recorder and while the process is running all i hear is buzzing sound, and also minor annoyance is that for capturing sinks it requires monitors. None of these issues are with pipewire backend, default audio selection will also automatically switch when changing default device during recording and it also supports capturing single application. So I can't really imagine anyone wanting that instead of proper working audio capture.

Reply:
Are you saying that wf-recorder current git without any patching has completely broken pulseaudio support? This would be quite surprising to me since 1) users have not reported such problems with pulseaudio recording and 2) I run pipewire with default wf-recorder git and there is no such issues. This makes me wonder if your patches break pulseaudio support in the mix.

Reply:
Yeah, of course I mean unpatched, can also reproduce on wf-recorder from Arch repo. May just be my system then.

> This makes me wonder if your patches break pulseaudio support in the mix.

You can try to build from this branch :)

Reply:
Yes, probably your system. But it would be nice to have a runtime option to select pulseaudio instead of pipewire if both are built in and available. It's not as easy if you have to rebuild wf-recorder with `-Dpipewire=disabled`, just to force pulseaudio. Lastly, I don't think it's fair to say that users don't have a preference in this context. The other thing I notice is that it is more likely that pipewire causes wf-recorder to fail to build. First thing I notice is that there is no version requirement in the meson.build for pipewire dependency. Here, it fails to build here with `/usr/include/pipewire-0.3/pipewire/properties.h:15:10: fatal error: spa/utils/string.h: No such file or directory` (despite /usr/include/spa-0.2/spa/utils/string.h presence) and while this might be some pipewire fail, I think pipewire should be disabled by default.

Reply:
Can you post full build log? `libspa-0.2` is dependency of `libpipewire-0.3` so it should get picked up automatically.

Reply:
> If user is running pipewire I'd say there is no reason to choose the pulse backend.

my .2 cents as a user is that I would be surprised if wf-recorder changed some important default without bumping major version (at least).

Also, I'd be happy to use the pipewire backend and remove the PulseAudio replacement daemon, if pipewire worked flawlessy OOTB (which is not my case unfortunately)

Reply:
> if pipewire worked flawlessy OOTB

Does this PR work for you or not? Otherwise I don't see how that's relevant here.

Reply:
(sorry for jumping into this thread of comments)

what I meant is that I'd love to switch to the Pipewire backend without looking back but currently that's for me not possible, so I assume also other users like me run both Pulse and Pipewire at the same time. Having `wf-recorder` switch to the Pipewire backend could be a breaking change. Happy to helpt testing this patch before merge, though!

In short: please avoid breaking changes if possible